File: app/src/main/java/com/iutcalendar/settings/AlarmConditionFragment.java
package com.iutcalendar.settings;

import android.os.Bundle;
import android.util.Log;
import android.view.*;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.calendar.iutcalendar.R;
import com.iutcalendar.alarm.AlarmRing;
import com.iutcalendar.alarm.condition.AlarmConditionManager;
import com.iutcalendar.alarm.condition.AlarmConditionRecycleView;
import com.iutcalendar.alarm.condition.AlarmCondtion;
import com.iutcalendar.calendrier.DateCalendrier;
import com.iutcalendar.dialog.DialogMessage;

public class AlarmConditionFragment extends Fragment {

    RecyclerView recyclerViewConstraint;
    View view;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_alarm_condition, container, false);

        initVaraible();


        updateConditions();
        return view;
    }

    private void initVaraible() {
        recyclerViewConstraint = view.findViewById(R.id.recycleView);
    }

    private void updateConditions() {
        AlarmConditionRecycleView adapter = new AlarmConditionRecycleView(getContext(),
                AlarmConditionManager.getInstance(getContext()).getAllConditions(), this::saveConditions, this::updateConditions);
        recyclerViewConstraint.setAdapter(adapter);
        recyclerViewConstraint.setLayoutManager(new LinearLayoutManager(getActivity()));

        //save si des changement de constraint on été fait
        saveConditions();
        Log.d("Constraint", "updateConstraint");
    }

    private void saveConditions() {
        AlarmConditionManager.getInstance(getContext()).save(getContext());
    }

    private void addCondition() {

        //demande heure début
        AlarmRing.askTime(getContext(), getString(R.string.BorneInf), (view, hourOfDayBeging, minuteBeging) -> {
            long beging = DateCalendrier.getHourInMillis(hourOfDayBeging, minuteBeging);

            //demande heure fin
            AlarmRing.askTime(getContext(), getString(R.string.BorneSup), (view1, hourOfDayEnd, minuteEnd) -> {
                long end = DateCalendrier.getHourInMillis(hourOfDayEnd, minuteEnd);

                if (beging > end) {
                    DialogMessage.showWarning(getContext(), getString(R.string.Interval), getString(R.string.born_sup_et_inf_inverser));
                    return;
                }


                //demande heure sonnerie
                AlarmRing.askTime(getContext(), getString(R.string.TimeToRing), (view2, hourOfDayAlarmAt, minuteAlarmAt) -> {
                    long alarmAt = DateCalendrier.getHourInMillis(hourOfDayAlarmAt, minuteAlarmAt);


                    AlarmConditionManager.getInstance(getContext()).addCondition(new AlarmCondtion(beging, end, alarmAt));
                    updateConditions();
                });
            });
        });
    }


    /*#################MENU BAR#################*/
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_action_constraint_alarm, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.addBtn) {
            addCondition();

        } else if (id == R.id.aideBtn) {
            DialogMessage.showAide(getContext(), getString(R.string.Help), getString(R.string.aide_conditon_alarm));
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/iutcalendar/settings/AlarmConditionFragment.kt
package com.iutcalendar.settings

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TimePicker
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.iutcalendar.alarm.AlarmRing
import com.iutcalendar.alarm.condition.AlarmCondition
import com.iutcalendar.alarm.condition.AlarmConditionManager
import com.iutcalendar.alarm.condition.AlarmConditionRecycleView
import com.iutcalendar.calendrier.DateCalendrier
import com.iutcalendar.dialog.DialogMessage
import com.iutcalendar.menu.AbstractFragmentWitheMenu
import com.univlyon1.tools.agenda.R

class AlarmConditionFragment : AbstractFragmentWitheMenu() {
    private var recyclerViewConstraint: RecyclerView? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        requireActivity().invalidateOptionsMenu()
        val view = inflater.inflate(R.layout.fragment_alarm_condition, container, false)
        initVariable(view)
        updateConditions()
        return view
    }

    private fun initVariable(view: View) {
        recyclerViewConstraint = view.findViewById(R.id.recycleView)
    }

    private fun updateConditions() {
        val adapter = AlarmConditionRecycleView(
            context,
            AlarmConditionManager.getInstance(context).allConditions,
            { itemPos: Int ->
                recyclerViewConstraint?.adapter?.notifyItemChanged(itemPos)
                saveConditions()
            }, { itemPos: Int ->
                recyclerViewConstraint?.adapter?.notifyItemRemoved(itemPos)
                recyclerViewConstraint?.adapter?.notifyItemRangeChanged(
                    itemPos,
                    AlarmConditionManager.getInstance(context).allConditions?.size!!
                )
                saveConditions()
            })
        { saveConditions() }
        recyclerViewConstraint?.let {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(activity)
        }

        //save si des changements de constraint ont été fait
        saveConditions()
        Log.d("Constraint", "updateConstraint")
    }

    private fun saveConditions() {
        AlarmConditionManager.getInstance(context).save(context)
    }

    private fun addCondition() {

        //demande heure début
        AlarmRing.askTime(
            context,
            getString(R.string.BorneInf)
        ) { _: TimePicker?, hourOfDayBegin: Int, minuteBegin: Int ->
            val begin: Long = DateCalendrier.getHourInMillis(hourOfDayBegin, minuteBegin)

            //demande heure fin
            AlarmRing.askTime(
                context,
                getString(R.string.BorneSup)
            ) { _: TimePicker?, hourOfDayEnd: Int, minuteEnd: Int ->
                val end: Long = DateCalendrier.getHourInMillis(hourOfDayEnd, minuteEnd)
                if (begin > end) {
                    DialogMessage.showWarning(
                        context,
                        getString(R.string.Interval),
                        getString(R.string.born_sup_et_inf_inverser)
                    )
                } else {
                    //demande heure sonnerie
                    AlarmRing.askTime(
                        context,
                        getString(R.string.TimeToRing)
                    ) { _: TimePicker?, hourOfDayAlarmAt: Int, minuteAlarmAt: Int ->
                        val alarmAt: Long =
                            DateCalendrier.getHourInMillis(hourOfDayAlarmAt, minuteAlarmAt)
                        AlarmConditionManager.getInstance(context)
                            .addCondition(AlarmCondition(begin, end, alarmAt))
                        updateConditions()
                    }
                }
            }
        }
    }

    /*#################MENU BAR#################*/
    override fun clickMenu(item: MenuItem) {
        val id = item.itemId
        Log.d("MenuBar", "fragment")
        if (id == R.id.addBtn) {
            addCondition()
        } else if (id == R.id.aideBtn) {
            DialogMessage.showAide(
                context,
                getString(R.string.Help),
                getString(R.string.aide_conditon_alarm)
            )
        }
    }
}